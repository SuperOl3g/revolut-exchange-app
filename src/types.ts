import React from 'react';
import { ActionType } from './actions/ActionTypes';

export type TCurrency = 'RUB' | 'USD' | 'EUR' | 'GBP';

export interface IFieldCallback {
  (
    e: React.SyntheticEvent<HTMLElement> | null,
    obj: {
      value: string;
      name: string;
    }
  ): void;
}

export interface IAction {
  type: ActionType;
  payload: {
    [key: string]: any;
  };
}
