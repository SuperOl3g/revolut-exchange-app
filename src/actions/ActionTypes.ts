export enum ActionType {
  RATES_REQUEST = 'RATES_REQUEST',
  RATES_REQUEST_SUCCESS = 'RATES_REQUEST_SUCCESS',

  SOURCE_CURRENCY_CHANGE = 'SOURCE_CURRENCY_CHANGE',
  TARGET_CURRENCY_CHANGE = 'TARGET_CURRENCY_CHANGE'
}
