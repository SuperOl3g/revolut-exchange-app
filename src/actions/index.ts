export enum ActionType {
  RATES_REQUEST = 'RATES_REQUEST',
  RATES_REQUEST_SUCCESS = 'RATES_REQUEST_SUCCESS',

  SOURCE_CURRENCY_CHANGE = 'SOURCE_CURRENCY_CHANGE',
  TARGET_CURRENCY_CHANGE = 'TARGET_CURRENCY_CHANGE',

  POCKETS_EXCHANGE = 'POCKETS_EXCHANGE',

  ALERT_ADD = 'ALERT_ADD',
  ALERT_REMOVE = 'ALERT_REMOVE'
}
