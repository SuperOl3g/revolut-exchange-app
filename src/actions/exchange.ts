import { TCurrency } from '../types';
import { Dispatch } from 'redux';
import { ActionType } from './ActionTypes';

export default function exchange(
  sourceCurrency: TCurrency,
  sourceAmount: number,
  targetCurrency: TCurrency,
  targetAmount: number
) {
  return (dispatch: Dispatch) => {
    dispatch({
      type: ActionType.EXCHANGE,
      payload: {
        sourceCurrency,
        sourceAmount,
        targetCurrency,
        targetAmount
      }
    });
  };
}
