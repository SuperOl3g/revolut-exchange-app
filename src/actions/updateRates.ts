import { Dispatch } from 'redux';
import { EXCHANGE_API } from '../constants/common';
import { ActionType } from './ActionTypes';
import { IStoreState } from '../store';

export default function updateRates() {
  return (dispatch: Dispatch, getState: () => IStoreState) => {
    const { sourceCurrency } = getState().exchange;

    dispatch({
      type: ActionType.RATES_REQUEST,
      payload: { baseCurrency: sourceCurrency }
    });

    return fetch(`${EXCHANGE_API}/latest?base=${sourceCurrency}`)
      .then(response => response.json())
      .then(({ rates }) => {
        const { RUB, EUR, USD, GBP } = rates;

        return { RUB, EUR, USD, GBP };
      })
      .then(rates =>
        dispatch({
          type: ActionType.RATES_REQUEST_SUCCESS,
          payload: { rates, baseCurrency: sourceCurrency }
        })
      );
  };
}
