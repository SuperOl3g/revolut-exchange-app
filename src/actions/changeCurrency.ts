import { ActionType } from './ActionTypes';
import { TCurrency } from '../types';
import { Dispatch, IStoreState } from '../store';
import { FieldType } from '../components/ExchangeWidget/WidgetBlock/WidgetBlock';
import updateRates from './updateRates';
import { DEFAULT_CURRENCY_ORDER } from '../constants/common';

function getNextCurrency(currency: TCurrency): TCurrency {
  return DEFAULT_CURRENCY_ORDER[
    (DEFAULT_CURRENCY_ORDER.indexOf(currency) + 1) %
      DEFAULT_CURRENCY_ORDER.length
  ];
}

export default function changeCurrency(type: FieldType, currency: TCurrency) {
  return (dispatch: Dispatch, getState: () => IStoreState) => {
    dispatch({
      type:
        type === FieldType.Source
          ? ActionType.SOURCE_CURRENCY_CHANGE
          : ActionType.TARGET_CURRENCY_CHANGE,
      payload: { currency }
    });

    dispatch(updateRates());

    const state = getState();
    const oppositeCurrency =
      state.exchange[
        type === FieldType.Source ? 'targetCurrency' : 'sourceCurrency'
      ];

    if (oppositeCurrency === currency) {
      dispatch({
        type:
          type === FieldType.Source
            ? ActionType.TARGET_CURRENCY_CHANGE
            : ActionType.SOURCE_CURRENCY_CHANGE,
        payload: { currency: getNextCurrency(oppositeCurrency) }
      });
    }
  };
}
