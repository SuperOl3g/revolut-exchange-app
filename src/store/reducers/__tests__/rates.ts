import reducer, { initState } from '../rates';
import { ActionType } from '../../../actions';
import { IAction } from '../../../types';

describe('Rates reducer', () => {
  it('should return the initial state', () => {
    expect(reducer(undefined, {} as IAction)).toMatchSnapshot();
  });

  it('should handle RATES_REQUEST_SUCCESS', () => {
    expect(
      reducer(initState, {
        type: ActionType.RATES_REQUEST_SUCCESS,
        payload: {
          rates: {
            RUB: 1,
            EUR: 0.0141689219,
            USD: 0.0156382391,
            GBP: 0.0129149723
          },
          baseCurrency: 'RUB'
        }
      })
    ).toMatchSnapshot();
  });
});
