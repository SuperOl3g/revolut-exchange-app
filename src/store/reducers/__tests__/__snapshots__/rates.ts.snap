// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rates reducer should handle RATES_REQUEST_SUCCESS 1`] = `
Object {
  "RUB": Object {
    "EUR": 0.0141689219,
    "GBP": 0.0129149723,
    "RUB": 1,
    "USD": 0.0156382391,
  },
}
`;

exports[`Rates reducer should return the initial state 1`] = `Object {}`;
